R/fg_pip.R

```diff
@@ -30,61 +30,86 @@ fg_pip <- function(country,
+  #NEW: iterate over survey files
+  for (svy_id in seq_along(unique_survey_files)) {
 
-    svy_data <- get_svy_data(tmp_metadata[["cache_id"]],
-      reporting_level = tmp_metadata[["pop_data_level"]],
-      path = tmp_metadata$path
+    # Extract country-years for which stats will be computed from the same files
+    # tmp_metadata <- interpolation_list[[unique_survey_files[svy_id]]]$tmp_metadata
+    svy_data <- get_svy_data(interpolation_list[[unique_survey_files[svy_id]]]$cache_ids,
+                             reporting_level = interpolation_list[[unique_survey_files[svy_id]]]$reporting_level,
+                             path = interpolation_list[[unique_survey_files[svy_id]]]$paths
     )
 
-    # Compute estimated statistics using the fill_gap method
-    if (debug) debugonce(wbpip:::prod_fg_compute_pip_stats)
-    tmp_stats <- wbpip:::prod_fg_compute_pip_stats(
-      request_year = ctry_years[["reporting_year"]][i],
-      data = svy_data,
-      predicted_request_mean = tmp_metadata[["predicted_mean_ppp"]],
-      svy_mean_lcu = tmp_metadata[["survey_mean_lcu"]],
-      survey_year = tmp_metadata[["survey_year"]],
-      default_ppp = tmp_metadata[["ppp"]],
-      ppp = ppp,
-      distribution_type = tmp_metadata[["distribution_type"]],
-      poverty_line = povline,
-      popshare = popshare
-    )
+    # Extract unique combinations of country-year
+    ctry_years <- subset_lkup(country = country,
+                              year = year,
+                              welfare_type = welfare_type,
+                              reporting_level = reporting_level,
+                              lkup = interpolation_list[[unique_survey_files[svy_id]]]$ctry_years)
+
+    results_subset <- vector(mode = "list", length = nrow(ctry_years))
+
+    for (ctry_year_id in seq_along(ctry_years$interpolation_id)) {
+
+      # Extract records to be used for a single country-year estimation
+      tmp_metadata <- metadata[metadata$interpolation_id == ctry_years[["interpolation_id"]][ctry_year_id], ]
+
+      # Compute estimated statistics using the fill_gap method
+      if (debug) debugonce(wbpip:::prod_fg_compute_pip_stats)
+      tmp_stats <- wbpip:::prod_fg_compute_pip_stats(
+        request_year = ctry_years[["reporting_year"]][ctry_year_id],
+        data = svy_data,
+        predicted_request_mean = tmp_metadata[["predicted_mean_ppp"]],
+        svy_mean_lcu = tmp_metadata[["survey_mean_lcu"]],
+        survey_year = tmp_metadata[["survey_year"]],
+        default_ppp = tmp_metadata[["ppp"]],
+        ppp = ppp,
+        distribution_type = tmp_metadata[["distribution_type"]],
+        poverty_line = povline,
+        popshare = popshare
+      )
+
+      # Ensure that tmp_metadata has a single row
+      vars_to_collapse <- c(
+        "survey_id", "cache_id", "surveyid_year", "survey_year",
+        "survey_acronym", "survey_coverage", "survey_comparability",
+        "comparable_spell", "welfare_type", "distribution_type",
```

Comment:
This could be moved outside the loop. The `vars_to_collapse` should be the same for every iteration.

Reply:
This has been moved out of the loop